### PR TITLE
As a user I can query the metrics of an application

### PR DESCRIPTION
**Pull request by qibobo**

### CODE REVIEW

File: api/lib/routes/metricHelper.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    parameters.endTime = -1;
+  } else {
+    endTime = parseInt(endTime);
+    if (!routeHelper.isInteger(endTime)) {
+      return { valid: false, message: "end-time must be an integer" };
+    } else {
+      parameters.endTime = endTime;
+    }
+  }
+  if (routeHelper.isMissing(order)) {
```
Comment (boyang9527) [inline api/lib/routes/metricHelper.js]: isn't the default ASC?

File: api/lib/routes/metricHelper.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    parameters.order = DESC;
+  } else {
+    if (!routeHelper.isString(order)) {
+      return { valid: false, message: "order must be a string" };
+    } else {
+      order = order.toLowerCase();
+      if (order !== DESC && order !== ASC) {
```
Comment (boyang9527) [inline api/lib/routes/metricHelper.js]: you convert to lower case, but compare with upper case

File: api/lib/routes/metricHelper.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    if (!routeHelper.isInteger(page)) {
+      return { valid: false, message: "page must be an integer" };
+    }
+    if (page <= 0) {
+      return { valid: false, message: "page must be greater than 0" };
+    }
+    parameters.page = page;
+  }
+
+  if (routeHelper.isMissing(resultsPerPage)) {
```
Comment (boyang9527) [inline api/lib/routes/metricHelper.js]: I am assuming if resultsPerPage is not specified, we return all

File: api/test/unit/routes/metricHelper.test.js [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+          validateResult = helper.parseParameter(requestObj);
+          expect(validateResult.valid).to.equal(false);
+          expect(validateResult.message).to.equal("metric-type must be a string");
+        });
+      });
+    });
+
+    context("validate start-time", function() {
+      context("start-time is undefined", function() {
+        it("valid is true, startTime is set to 0", function() {
```
Comment (boyang9527) [inline api/test/unit/routes/metricHelper.test.js]: change "valid is true" to "returns true"

File: api/lib/routes/metrics.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+module.exports = function(settings) {
+  var express = require("express");
+  var router = express.Router();
+  var logger = require("../log/logger");
+  var HttpStatus = require("http-status-codes");
+  var metricsCollectorUtil = require("../utils/metricsCollectorUtils")(settings.metricsCollector);
+  var metricHelper = require("./metricHelper");
+  var routeHelper = require("./routeHelper");
+
+  router.get("/:app_id/metrics", function(req, resp) {
```
Comment (boyang9527) [inline api/lib/routes/metrics.js]: metricType should be in the path

File: api/lib/routes/metricHelper.js (hunk 2 above)
Comment (qibobo) [inline api/lib/routes/metricHelper.js]: @boyang9527 DESC and ASC are variable names, their values are "asc" and "desc".

File: api/lib/routes/metricHelper.js (hunk 3 above)
Comment (qibobo) [inline api/lib/routes/metricHelper.js]: @boyang9527  for scaling history, we set a default value 10 if resultPerpage is missing, so here you mean we change the rule just for metrics or apply it to scaling history, too?

File: api/lib/routes/metrics.js (hunk 5 above)
Comment (qibobo) [inline api/lib/routes/metrics.js]: @boyang9527 in our api document, metricType is a query parameter. Here need your confirm we set it to path variable.

File: api/lib/routes/metricHelper.js (hunk 3 above)
Comment (qibobo) [inline api/lib/routes/metricHelper.js]: @boyang9527   the default of cf is 50, so set to 50;

File: api/lib/routes/metrics.js (hunk 5 above)
Comment (qibobo) [inline api/lib/routes/metrics.js]: @boyang9527 fixex.

File: api/test/unit/routes/metricHelper.test.js (hunk 4 above)
Comment (qibobo) [inline api/test/unit/routes/metricHelper.test.js]: fixed.

File: api/lib/routes/metricHelper.js (hunk 2 above)
Comment (boyang9527) [inline api/lib/routes/metricHelper.js]: got u... a bit confused... you may want to change to upper case

File: api/lib/routes/metricHelper.js (hunk 3 above)
Comment (boyang9527) [inline api/lib/routes/metricHelper.js]: I am good to follow the cf way to make it consistent

File: api/lib/routes/metricHelper.js (hunk 2 above)
Comment (qibobo) [inline api/lib/routes/metricHelper.js]: @boyang9527 fixed.

**Comment by boyang9527** [comment]
LGTM

**Comment by boyang9527** [comment]
@pradyutsarma  @paltanmoy please help take some time to review this PR.. 
I understand it is a big PR,  while test code to verify settings and parameters takes the major part.

**Comment by paltanmoy** [comment]
@qibobo Integration test fails in Travis. Could you please have a look ?
# integration_test
src/integration/integration_api_metricscollector_test.go:37: not enough arguments in call to fakeMetricsPolling
src/integration/integration_api_metricscollector_test.go:73: undefined: models.MetricNameMemory
src/integration/integration_api_metricscollector_test.go:106: undefined: models.MetricNameMemory
src/integration/integration_api_metricscollector_test.go:124: undefined: models.MetricNameMemory
src/integration/integration_api_metricscollector_test.go:133: undefined: models.MetricNameMemory
src/integration/integration_api_metricscollector_test.go:153: undefined: models.MetricNameMemory
src/integration/integration_api_metricscollector_test.go:162: undefined: models.MetricNameMemory
src/integration/integration_api_metricscollector_test.go:182: undefined: models.MetricNameMemory

**Comment by qibobo** [comment]
@paltanmoy fixed.

**Comment by qibobo** [comment]
@paltanmoy @pradyutsarma  The code has been updated and all check passed. Do you have more comments?

**Comment by qibobo** [comment]
@boyang9527 @paltanmoy we plan to get this pr done today, please give your comments if there is, or we will get it merged. thx.

**Comment by paltanmoy** [comment]
LGTM

**Pull request merged**